returnn/datasets/postprocessing.py

```diff
@@ -0,0 +1,239 @@
+
+        if self._map_outputs is not None:
+            tdict = TensorDict(self._map_outputs)
+            self.num_outputs = {
+                k: (t.sparse_dim.size if t.sparse_dim else t.shape[-1] or 1, t.ndim) for k, t in tdict.data.items()
+            }
+            self.num_inputs = self.num_outputs[self._default_input][0]
+        else:
+            self.num_inputs = dataset.num_inputs
+            self.num_outputs = dataset.num_outputs
+
+    def init_seq_order(self, epoch: Optional[int] = None, seq_list=None, seq_order=None):
+        """
+        :param epoch:
+        :param seq_list:
+        :param seq_order:
+        :return: whether the order changed (True is always safe to return)
+        """
+        super().init_seq_order(epoch=epoch, seq_list=seq_list, seq_order=seq_order)
+
+        if epoch is None:
+            self._num_seqs = 0
+            return True
+
+        if self._dataset is None:
+            self._dataset = init_dataset(self._dataset_def, parent_dataset=self)
+        self._dataset.init_seq_order(epoch=epoch, seq_list=seq_list, seq_order=seq_order)
+        self._data_iter = enumerate(self._build_dataset_iter(self._dataset))
+        self._data_iter_seq_idx = -1
+        self._lazy_init_num_outputs()
+        return True
+
+    def _collect_single_seq(self, seq_idx: int) -> Optional[DatasetSeq]:
+        assert seq_idx > self._data_iter_seq_idx, "collecting seqs must be done strictly monotonically"
+        self._data_iter_seq_idx = seq_idx
+
+        while True:
+            try:
+                loaded_seq_idx, tensor_dict = next(self._data_iter)
+            except StopIteration:
+                return None
+            if loaded_seq_idx != seq_idx:
+                continue
+            seq = DatasetSeq(features=tensor_dict.as_raw_tensor_dict(), seq_idx=seq_idx)
+            return seq
+
+    def _build_dataset_iter(self, dataset: Dataset) -> Iterator[TensorDict]:
+        """
+        :return: an iterator applying both the segment level and across-segment transformations on the given dataset
+        """
+        return self._map_seq_stream(map(self._map_seq, self._iterate_dataset(dataset)))
+
+    def _data_dict_to_tensor_dict(
+        self, dataset: Dataset, data_dict: Dict[str, ndarray], seq_len: NumbersDict
+    ) -> TensorDict:
+        """
+        :return: the given data dict converted to a TensorDict class
+        """
+
+        def _make_tensor(name: str, data: ndarray) -> Tensor:
+            dims = None
+            sparse_dim = None
+            dtype = data.dtype
+            if data.dtype.name.startswith("str"):
+                # TODO: is this correct?
+                dims = []
+                dtype = "string"
+            if dims is None:
+                dims, sparse_dim = self._dim_cache.get(name, (None, None))
+            if dims is None:
+                dims = [Dim(dimension=seq_len[name], name=f"{name}_num_frames")]
```

Comment:
And how we'd deal with seq lengths?

Reply:
If it is dynamic, it is `dimension=None`.